src/index.ts

```diff
@@ -1,39 +1,59 @@
-import { domainCache } from './domainCache';
-import { hashDomain } from './hashDomain';
-import { MirrorNode, NetworkType } from './mirrorNode';
-import { PollingTopicSubscriber } from './topicSubscriber/pollingTopicSubscriber';
-import { NameHash } from './types/NameHash';
-import { SecondLevelDomain } from './types/SecondLevelDomain';
-import { TopLevelDomain } from './types/TopLevelDomain';
-
-export const TEST_TLD_TOPIC_ID = '0.0.48097305';
-export const MAIN_TLD_TOPIC_ID = '0.0.1234189';
-
+import { hashDomain } from "./hashDomain";
+import { MemoryCache } from "./MemoryCache";
+import { MirrorNode, NetworkType } from "./mirrorNode";
+import { PollingTopicSubscriber } from "./topicSubscriber/pollingTopicSubscriber";
+import { ICache } from "./types/Cache";
+import { NameHash } from "./types/NameHash";
+import { SecondLevelDomain } from "./types/SecondLevelDomain";
+import { TopLevelDomain } from "./types/TopLevelDomain";
+
+export const TEST_TLD_TOPIC_ID = "0.0.48097305";
+export const MAIN_TLD_TOPIC_ID = "0.0.1234189";
+
+export {
+  ICache,
+  Links,
+  MessageObject,
+  MessagesResponse,
+  NFT,
+  NFTsResponse,
+  NameHash,
+  SecondLevelDomain,
+  TopLevelDomain,
+} from "./types";
 
 export class Resolver {
   mirrorNode: MirrorNode;
-
   private _isCaughtUpWithTopic = new Map<string, boolean>();
   private _subscriptions: (() => void)[] = [];
+  private cache: ICache;
 
-  constructor(networkType: NetworkType, authKey = '') {
+  constructor(networkType: NetworkType, authKey = "", cache?: ICache) {
     this.mirrorNode = new MirrorNode(networkType, authKey);
+    if (!cache) {
+      this.cache = new MemoryCache();
+    } else {
+      this.cache = cache;
+    }
   }
 
   /**
    * @description Initializes all topic subscriptions.
    */
   public init() {
     this.getTopLevelDomains().then(() => {
-      const knownTlds = Array.from(domainCache.tld.values());
-      knownTlds.forEach(tld => {
-        this.getSecondLevelDomains(tld.topicId);
-      })
+      Promise.resolve(this.cache.getTlds()).then((knownTlds) => {
```

Comment:
Wrapping `this.cache.getTlds()` in a promise will be redundant if the cache always returns a Promise.